src/environment.c

```diff
@@ -0,0 +1,137 @@
+#include <sys/queue.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <stdio.h>
+#include "environment.h"
+
+static SLIST_HEAD( csh_env_llist, csh_env_entry ) csh_env = SLIST_HEAD_INITIALIZER(csh_env);
+struct csh_env_entry {
+    char *name;
+    char *value;
+    SLIST_ENTRY(csh_env_entry) list;
+};
+
+static struct csh_env_entry *csh_get_env_entry(const char *name) {
+    struct csh_env_entry *cur;
+    SLIST_FOREACH(cur, &csh_env, list) {
+        if(strcmp(name, cur->name) == 0) {
+            return cur;
+        }
+    }
+    return NULL;
+}
+
+ __attribute__ ((malloc)) static char *csh_copy_char_and_grow(char *dest, size_t at, size_t *dest_len, char c) {
```

Comment:
I like the use of memory safety attributes. The semantic of __attribute__((malloc)) differs depending on which version of GCC you use:

https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html
_The malloc attribute is used to tell the compiler that a function may be treated as if any non-NULL pointer it returns cannot alias any other pointer valid when the function returns and that the memory has undefined content. This will often improve optimization. Standard functions with this property include malloc and calloc. realloc-like functions do not have this property as the memory pointed to does not have undefined content._

https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Common-Function-Attributes.html
_Attribute malloc indicates that a function is malloc-like, i.e., that the pointer P returned by the function cannot alias any other pointer valid when the function returns, and moreover no pointers to valid objects occur in any storage addressed by P. In addition, GCC predicts that a function with the attribute returns non-null in most cases._ 

It seems to me that csh_copy_char_and_grow() is a "realloc()-like" function, which is not intended for older version of __attribute__((malloc)), this is no big deal however.

Additionally, __attribute__((malloc(<deallocator>))) may be used to indicate that the provided deallocator (free() included) should be called, this also indicates to the caller that they are now responsible for the memory.
If we are using __attribute__s anyway, this deallocator version may be fitting too.

Reply:
Given that the last 4.x gcc series was release in 2016 and that we are using GCC > =11 both on PC and target, we can safely assume that the 2nd definition applies here. I'll update to use `__attribute__ ((malloc (free)))`, will that work ?

Reply:
Actually, I may have misunderstood: seems like realloc and friends should not use this attribute as the memory returned by this type of function is, at least partially, initialized. Given the function does not itself allocate the initial memory, I think the better approach is simply to remove the attribute altogether. Thougths ?

Reply:
I believe the semantics of `__attribute__((malloc))` and `__attribute__((malloc(free)))` are entirely different.
So I think `__attribute__((malloc()))` can still be applied to realloc-like functions. I also like this `__attribute__` as a compiler-promise, that the caller is responsible for the memory. Although I've had little success having it issue warnings in PyCSH, but that might've been because I used the wrong compiler warning arguments.
I suppose it's also possible that  `__attribute__((malloc()))` will be improved in future versions of GCC, to better track memory management. In which case it might be worth to introduce it now, even if it doesn't function optimally yet.

Reply:
For reference, I use `__attribute__((malloc()))` for a couple of different `.__new__()` methods in PyCSH, but it might be too difficult for GCC to track the returned pointer through the Python interpreter's reference counting :sweat_smile: 

__attribute__((malloc, malloc(SlashCommand_dealloc, 1)))
static PyObject * SlashCommand_new(PyTypeObject *type, PyObject *args, PyObject *kwds);

Reply:
As agreed, I've removed the attribute in commit 95b6e81aec86a19edeeb6d3916e0c99401b5f104